Source/Libraries/openXDA.Model/Channels/ChannelData.cs

```diff
@@ -415,6 +501,93 @@ private static List<Tuple<int, List<DataPoint>>> Decompress_Legacy(byte[] data)
             return result;
         }
 
+        /// <summary>
+        /// Decompresses a Digital stored as compresed series of changes
+        /// </summary>
+        /// <param name="data"> The compressed <see cref="byte[]"/></param>
+        /// <returns> a Dictionary mapping a SeriesID to a decopmressed <see cref="List{DataPoint}"/></returns>
+        private static List<Tuple<int, List<DataPoint>>> Decompress_Digital(byte[] data)
+        {
+            List<Tuple<int, List<DataPoint>>> result = new List<Tuple<int, List<DataPoint>>>();
+            byte[] uncompressedData;
+            int offset;
+            int m_samples;
+            DateTime startTime;
+            int seriesID;
+            long samplingrate;
+            List<DataPoint> points = new List<DataPoint>();
+
+            // Restore the GZip header before uncompressing
+            data[0] = legacyHeader[0];
+            data[1] = legacyHeader[1];
+
+            uncompressedData = GZipStream.UncompressBuffer(data);
+            offset = 0;
+
+            m_samples = LittleEndian.ToInt32(uncompressedData, offset);
+            offset += sizeof(int);
+
+            if (m_samples < 2)
+                throw new InvalidOperationException("Digital Data must have the first and last point encoded.");
+
+            seriesID = LittleEndian.ToInt32(uncompressedData, offset);
+            offset += sizeof(int);
+
+            startTime = new DateTime(LittleEndian.ToInt64(uncompressedData, offset));
+            offset += sizeof(long);
+
+            samplingrate = LittleEndian.ToInt64(uncompressedData, offset);
+            offset += sizeof(long);
+
+            double decompressionOffset = LittleEndian.ToDouble(uncompressedData, offset);
+            double decompressionScale = LittleEndian.ToDouble(uncompressedData, offset + sizeof(double));
+            offset += 2 * sizeof(double);
+
+            // Always have the first point available.
+            int time = LittleEndian.ToUInt16(uncompressedData, offset);
+            offset += sizeof(int);
+            ushort compressedValue = LittleEndian.ToUInt16(uncompressedData, offset);
+            offset += sizeof(ushort);
+
+            double decompressedValue = decompressionScale * compressedValue + decompressionOffset;
+
+            double lastValue = decompressedValue;
+            long lastTime = time;
+
+            for (int i = 1; i < m_samples; i++)
+            {
+                time = LittleEndian.ToUInt16(uncompressedData, offset);
+                offset += sizeof(int);
+                compressedValue = LittleEndian.ToUInt16(uncompressedData, offset);
+                offset += sizeof(ushort);
+
+                decompressedValue = decompressionScale * compressedValue + decompressionOffset;
+                while (time > (lastTime + samplingrate))
+                {
+                    lastTime += samplingrate;
```

Comment:
I wonder if rounding errors could potentially introduce additional data points between state changes. Do you think it might cause problems downstream if `points.Count > m_samples`?

Reply:
`m_samples` get's removed from the blob so I am not sure that even matters.

By definition we assume these "Digitals" are evenly sampled so it will generate an evenly sampled flat line until it hits a state change. In theory it is possible that we Loose some points in between or generate extra ones, but since it's a flat line anyway I am not too concerned about that.

Occasionally we may compress a latched analog this way, which I suppose could be an issue, but I believe we already assume fixed sampling rate in a few places when doing math so not sure that really matters either.

Reply:
Off the top of my head, my main concern would be this.

https://github.com/GridProtectionAlliance/openXDA/blob/f0ccd33a6e4ae1ca74a5037b69001bcefada7ea1/Source/Libraries/FaultData/DataAnalysis/DataGroup.cs#L323-L331

If the actual number of decoded data points doesn't match `DataGroup.m_samples`, then the `DataGroup.Add()` method will fail to even include the digital series in the `DataGroup`. This is called directly by `DataGroup.FromData()` so the decoded data would just be silently dropped on the floor, and we would be wondering why it's not getting returned to the visualization.

Also, I suppose I'm assuming that there will be a follow-up PR that includes changes to `DataGroup.FromData()`?